### PR TITLE
Feat(eslint-plugin-fiori-tools): no comma in section title rule

### PR DESCRIPTION
**Pull request by AlinaGovoruhina**
## Description

This PR introduces a new ESLint rule **`sap-no-comma-in-section-title`** for the `@sap-ux/eslint-plugin-fiori-tools` package. The rule prevents commas in section and subsection titles on SAP Fiori Elements object pages, as commas are used as delimiters for backend message grouping and their presence in titles can cause incorrect message routing.

### Key changes:

**New rule (`sap-no-comma-in-section-title`)**
- Validates `UI.Facets` annotations on object pages for direct string labels containing commas
- Uses a **two-pass design**: annotation pass checks direct strings; i18n pass checks `.properties` file values whose keys are used as section labels via bindings like `{@i18n>key}`
- Covers both `ReferenceFacet` and nested `CollectionFacet` subsections
- Supports both OData V2 and V4 (XML and CDS)

**i18n infrastructure (new)**
- Added `FioriI18nSourceCode`, `I18nDocument`, `I18nEntry` AST types and parser for `.properties` files
- Added `I18nVisitNodeStep` traversal step and `FioriI18nRuleDefinition` type
- Extended `FioriLanguage` to parse and handle `.properties` files alongside XML, JSON, CDS, and `.change` files
- Extended `createFioriRule` factory with a `createI18n` hook for i18n-specific visitors

**Project context updates**
- `ParsedApp` now includes `i18nBundles: I18nBundle[]` populated from `sap.app.i18n` and `sap.ui5.models` ResourceModel entries in `manifest.json`
- `ApplicationParser` can reparse `.properties` files and update bundles incrementally
- Added `ProjectContext.resetForTesting()` for clean test isolation

**Documentation & configuration**
- Added rule documentation at `docs/rules/sap-no-comma-in-section-title.md`
- Rule registered in the plugin's recommended config and README rules table
- `.agents/skills` reference updated with the two-pass i18n design as the canonical example

## Type of change
- [x] New feature (non-breaking change which adds a new feature)

## How have you tested?
- Unit tests added for the new rule covering valid and invalid cases (V4 direct label, V4 CollectionFacet subsection, V2 direct label, V4 i18n binding)
- Unit tests added for `FioriI18nSourceCode` (parsing, traversal, location, edge cases)
- Unit tests added for i18n bundle reparsing in `ApplicationParser`

## Checklist:
- [ ] The code conforms to the [general development principles](https://github.com/SAP/open-ux-tools/blob/main/docs/Guidelines.md)
- [ ] Supplied as many details as possible on this change
- [ ] The code is easy to read and maintainable by others
- [ ] Corresponding changes to the documentation has been done
- [ ] Already existing and new unit tests pass locally
- [ ] I have reviewed and addressed all Hyperspace bot findings (or explicitly explained dismissals)
- [ ] I have done an Agentic review

- [ ] 🔄 Regenerate and Update Summary

<details>
<summary>PR Bot Information</summary>

**Version:** `1.31.2`

- Summary Prompt: [Default Prompt](https://github.tools.sap/Code-Change-Intelligence/pr-bot/blob/main/src/services/llm/prompts/summary_instructions_prompt.md)
- Event Trigger: `pull_request.opened`
- LLM: `anthropic--claude-4.6-sonnet`
- File Content Strategy: Full file content
- Output Template: Repository PR Template
- Correlation ID: `aaa27470-a5d0-11f1-9458-0c26e5facf8e`
</details>

### CODE REVIEW

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts [hunk 1]
```diff
@@ -505,3 +547,71 @@ function getMinUI5Version(manifest: Manifest): MinUI5Version | undefined {
...
+ */
+function collectI18nBundles(paths: string[], fileCache: Map<string, string>): I18nBundle[] {
+    const bundles: I18nBundle[] = [];
+    for (const filePath of paths) {
+        const uri = pathToFileURL(filePath).toString();
+        try {
+            const content = fileCache.get(uri);
+            if (content !== undefined) {
+                bundles.push({ uri, entries: parseI18nProperties(content) });
+            }
```
[the remark this answers is not in the document]
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts]: Fixed

File: packages/eslint-plugin-fiori-tools/test/rules/sap-no-comma-in-section-title.test.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        createInvalidTest(
+            {
+                name: 'V4: i18n entry value has comma',
+                filename: V4_I18N_PATH,
+                code:
+                    V4_I18N_CONTENT +
+                    `
+#XFLD,50: Label for a section
+tableSection00=table, section, 00
+`,
```
[the remark this answers is not in the document]
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/rules/sap-no-comma-in-section-title.test.ts]: Fixed

File: packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    const labelPropValue = getLabelPropValue(record);
+    if (!labelPropValue) {
+        return;
+    }
+    const i18nKey = extractI18nKey(getAttrOrChildText(labelPropValue, Edm.String));
+    if (!i18nKey) {
+        return;
+    }
+    const existing = keyToPageNames.get(i18nKey) ?? [];
+    keyToPageNames.set(i18nKey, [...existing, pageName]);
```
[the remark this answers is not in the document]
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts]: Fixed

File: packages/eslint-plugin-fiori-tools/src/project-context/project-context.ts [hunk 4]
```diff
@@ -237,6 +238,16 @@ export class ProjectContext {
...
+
+    /**
+     * Clears all static caches so tests run in isolation regardless of execution order.
+     */
+    public static resetForTesting(): void {
+        this.fileCache.clear();
+        this.instanceCache.clear();
+        this.updateCache.clear();
+        this.appRoots.clear();
+    }
```
[the remark this answers is not in the document]
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/project-context.ts:251]: Added artifact cache clear

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
+# Ensure Section and Subsection Titles Do Not Contain Commas (`sap-no-comma-in-section-title`)
+
+## Rule Details
+
+### Why Was This Rule Introduced?
+
+In SAP Fiori Elements object pages, section and subsection titles must not contain commas. Commas are used as delimiters for backend message grouping, so a comma in a section title can cause messages to be routed incorrectly, leading to unexpected behavior in error and status message handling.
```
Comment (lfindlaysap) [inline packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md]:
```suggestion
Section and subsection titles must not contain commas in object pages for SAP Fiori elements applications. Commas are used as delimiters for back-end message grouping, so a comma in a section title can cause messages to be routed incorrectly, which leads to unexpected behavior in error and status message handling.
```

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+This rule checks both:
```
Comment (lfindlaysap) [inline packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md]:
```suggestion
This rule checks the following scenarios:
```

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+- Direct string labels in OData annotation XML files (e.g., `String="Products, Details"`)
```
Comment (lfindlaysap) [inline packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md]:
```suggestion
- Direct string labels in XML files for OData annotations, for example `String="Products, Details"`
```

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+- i18n property values whose keys are bound as section labels (e.g., a key used as `{@i18n>sectionTitle}` where the value contains a comma)
```
Comment (lfindlaysap) [inline packages/eslint-plugin-fiori-tools/docs/rules/sap-no-comma-in-section-title.md]:
```suggestion
- i18n property values whose keys are bound as section labels, for example a key used as `{@i18n>sectionTitle}`, where the value contains a comma
```

File: packages/eslint-plugin-fiori-tools/README.md [hunk 9]
```diff
@@ -120,6 +120,7 @@ npx --yes @sap-ux/create@latest convert eslint-config --help
 
 |   Since   | Rule | Description | Recommended | Recommended for S/4HANA |
 |:---------:|------|-------------|:-----------:|:-----------------------:|
+|  new      | [sap-no-comma-in-section-title](docs/rules/sap-no-comma-in-section-title.md) | Ensures that section and subsection titles in SAP Fiori Elements object pages do not contain commas. | | ✅ |
```
Comment (lfindlaysap) [inline packages/eslint-plugin-fiori-tools/README.md]:
```suggestion
|  new      | [sap-no-comma-in-section-title](docs/rules/sap-no-comma-in-section-title.md) | Ensures that section and subsection titles in object pages for SAP Fiori elements applications do not contain commas. | | ✅ |
```

**Review by lfindlaysap — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts [hunk 10]
```diff
@@ -505,3 +547,69 @@ function getMinUI5Version(manifest: Manifest): MinUI5Version | undefined {
...
+ * @param paths - Absolute filesystem paths to .properties files
+ * @param fileCache - Map of file URIs to their contents (may be a proxy that reads from disk)
+ * @returns Array of I18nBundle objects with parsed entries
+ */
+function collectI18nBundles(paths: string[], fileCache: Map<string, string>): I18nBundle[] {
+    const bundles: I18nBundle[] = [];
+    for (const filePath of paths) {
+        const uri = pathToFileURL(filePath).toString();
+        try {
+            const content = fileCache.get(uri) ?? readFileSync(fileURLToPath(uri), { encoding: 'utf8', flag: 'r' });
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts:608]: Why not using filePath directly?

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts [hunk 11]
```diff
@@ -505,3 +547,69 @@ function getMinUI5Version(manifest: Manifest): MinUI5Version | undefined {
...
 }
+
+/**
+ * Parses a .properties file content into a key/value map.
+ * Lines starting with # or ! are treated as comments and skipped.
+ *
+ * @param content - The raw text content of the .properties file
+ * @returns Record mapping i18n keys to their values
+ */
+function parseI18nProperties(content: string): Record<string, string> {
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts:558]: Don't we have similar logic in source-code.ts: parseI18nToAst?

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts [hunk 12]
```diff
@@ -505,3 +547,69 @@ function getMinUI5Version(manifest: Manifest): MinUI5Version | undefined {
...
+
+/**
+ * Collects unique i18n file paths declared in the manifest.
+ * Checks sap.app.i18n (string form) and sap.ui5.models ResourceModel entries.
+ *
+ * @param webappPath - The absolute path to the webapp directory
+ * @param manifest - The parsed manifest object
+ * @returns Deduplicated array of absolute filesystem paths to i18n files
+ */
+function getI18nFilePaths(webappPath: string, manifest: Manifest): string[] {
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts:581]: AI mentioned that the implementation of getI18nFilePaths is incomplete (appI18n can also be an object, check packages/project-access/src/project/i18n/i18n.ts getI18nAppPath).
Can we use getRelativeI18nPropertiesPaths from @sap-ux/project-access? (It may be exported in packages/project-access/src/project/i18n/index.ts)

File: packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+/**
+ * Checks a .properties file for i18n entries whose keys are used as section/subsection labels
+ * and whose values contain commas.
+ *
+ * @param sourceCode - The Fiori i18n source code being linted
+ * @returns Array of diagnostics for i18n entries with comma-containing values
+ */
+function checkI18nSource(sourceCode: FioriI18nSourceCode): NoCommaInSectionTitle[] {
+    const problems: NoCommaInSectionTitle[] = [];
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts:178]: Do we need a guard with app.i18nBundles here?

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts [hunk 14]
```diff
@@ -505,3 +547,69 @@ function getMinUI5Version(manifest: Manifest): MinUI5Version | undefined {
...
+function parseI18nProperties(content: string): Record<string, string> {
+    const entries: Record<string, string> = {};
+    for (const line of content.split('\n')) {
+        const trimmed = line.trim();
+        if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith('!')) {
+            continue;
+        }
+        const eqIdx = trimmed.indexOf('=');
+        if (eqIdx > 0) {
+            entries[trimmed.slice(0, eqIdx).trim()] = trimmed.slice(eqIdx + 1);
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts:567]: AI mentioned whitespace before the value should be ignored. So use trimmed.slice(eqIdx + 1).trimStart();
(https://github.com/UI5/openui5/blob/master/src/sap.ui.core/src/sap/base/util/Properties.js regexp for key/value in rEscapesOrSeparator)

**Review by marcovieth — COMMENTED** [review]
Please check the comments

File: packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts (hunk 13 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts:178]: Since it checks that sourceCode is an instanceof FioriI18nSourceCode first, I don't think it's necessary.

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts (hunk 14 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts]: Adjusted for leading or trailing whitespaces + added unit tests in https://github.com/SAP/open-ux-tools/pull/5140/commits/177300b944e957b5cba33feb6a17931142aca088

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts (hunk 10 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts]: Fixed

File: packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    for (const page of app.pages) {
+        if (page.type !== 'object-page') {
+            continue;
+        }
+        const entityType = page.entity?.structuredType;
+        if (!entityType) {
+            continue;
+        }
+        collectFacetI18nKeys(entityType, page.targetName, parsedService, keyToPageNames);
+    }
```
[the remark this answers is not in the document]
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts:155]: This is correct. Since a project level i18n.properties file can be used by multiple apps - so we want to check it for every app. But if done so, we get eslint results for all apps, so in case of  i18n files only the explicitly selected app has to be checked.

File: packages/eslint-plugin-fiori-tools/src/project-context/artifacts.ts [hunk 16]
```diff
@@ -11,14 +20,37 @@ import type { WorkerResult } from './types.js';
...
+            const appI18nPaths = await getI18nPropertiesPaths(app.manifestPath);
+            i18nPaths.push(appI18nPaths['sap.app']);
+            if (!isCap) {
+                for (const model of Object.values(appI18nPaths.models)) {
+                    if (!i18nPaths.includes(model.path)) {
+                        i18nPaths.push(model.path);
+                    }
+                }
+                i18nPathsByApp = { ...i18nPathsByApp, [app.appRoot]: i18nPaths };
+            }
```
[the remark this answers is not in the document]
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/artifacts.ts]: Updated

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts (hunk 12 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts]: Updated with methods from @sap-ux/i18n

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts (hunk 11 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts:578]: Reused this function here

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts [hunk 17]
```diff
@@ -277,23 +303,59 @@ export class ApplicationParser {
...
+        const app = index.apps[appUri];
+        if (!app) {
+            return;
+        }
+        const bundleIndex = app.i18nBundles.findIndex((bundle) => bundle.uri === uri);
+        if (bundleIndex < 0) {
+            return;
+        }
+        try {
+            const content = fileCache.get(uri) ?? readFileSync(appRoot, { encoding: 'utf8', flag: 'r' });
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts]: Isn't appRoot a directory and shouldn't we use fileURLToPath(uri) instead?

File: packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+ *
+ * @param projectContext - The project context
+ * @returns Map from i18n key to the list of page names using that key as a label
+ */
+export function collectSectionLabelKeys(projectContext: ProjectContext): Map<string, string[]> {
+    const keyToPageNames = new Map<string, string[]>();
+    const appUri = pathToFileURL(projectContext.index.appRoot).toString();
+    const app = projectContext.linkedModel.apps[appUri];
+    const parsedApp = projectContext.index.apps[appUri];
+    const parsedService = projectContext.getIndexedServiceForMainService(parsedApp);
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts:145]: Do we need some null-checks for app and parsedApp?

File: packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+                'Section or subsection title must not contain commas. Commas are used as delimiters for backend message grouping.'
+        }
+    },
+    check(context) {
+        if (
+            !(context.sourceCode instanceof FioriAnnotationSourceCode) &&
+            !(context.sourceCode instanceof FioriI18nSourceCode)
+        ) {
+            return [];
+        }
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts]: Isn't this check redundant?

Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts:217]: Isn't this check redundant?

**Review by marcovieth — COMMENTED** [review]
Please check my comments

File: packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts (hunk 19 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts:217]: In the config all rules apply to all files, so it makes sense to check the file is not .json https://github.com/SAP/open-ux-tools/blob/d65c450181adcfce74bcec5eeac14e0a43c9af1c/packages/eslint-plugin-fiori-tools/src/index.ts#L483
It could be a good idea to separate the config into rule groups based on which files they apply to, but better to explore this separately.

File: packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts (hunk 18 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/linker/i18n.ts:145]: Thanks, added

File: packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts (hunk 17 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/project-context/parser/parser.ts]: Replaced with file path

File: packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts (hunk 19 above)
Comment (marcovieth) [reply packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts]: Yes, we need an empty return [] for irrelevant files. But we already have it at the end of the check function.

What I mean is in the context with lines 218-224, which also check the same instanceof.
So either remove the initial check or
the third condition in line 221 which currently is always true, together with the final return [] which is never reached.

Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/rules/sap-no-comma-in-section-title.ts]: Ah, in this case it is not necessary, removed the block.

**Review by marcovieth — DISMISSED** [review]
looks good to me
